File: downloader.py
import praw
import requests
import os
import json
import traceback
from imgur_downloader import ImgurDownloader
import subprocess
from redgifs import API
import sys
api = API()
api.login()
from datetime import datetime, timezone


def urltolocalurl(url,post,name,output_dir,safe_title):
    try:
        if "reddit.com/gallery" in url:
            # This is a gallery post
            for i, media in enumerate(post.media_metadata.values()):
                media_url = media['s']['u']
                file_extension = os.path.splitext(media_url.split("?")[0])[1]  # Remove URL parameters
                file_path = os.path.join(f"{output_dir}/{name}", f'{safe_title}_{i}{file_extension}')
                if not os.path.exists(file_path):
                    response = requests.get(media_url)
                    with open(file_path, 'wb') as f:
                        f.write(response.content)
                        
            #images_data.append({'url': f"{output_dir}/{name}/{safe_title}_{i}{file_extension}", 'title': safe_title})    
            localurl=f"{output_dir}/{name}/{safe_title}_{i}{file_extension}"
                                
        elif "redgifs.com/watch" in url:
            # This is a redgifs post
            file_path = os.path.join(f"{output_dir}/{name}", f'{safe_title}.mp4')
            if not os.path.exists(file_path):
                result = api.get_gif(url.split("/")[-1])
                api.download(result.urls.hd, f'{output_dir}/{name}/{safe_title}.mp4')

                
            #images_data.append({'url': f'{output_dir}/{name}/{safe_title}.mp4'.replace("\\", "/"), 'title': safe_title})
            localurl = f'{output_dir}/{name}/{safe_title}.mp4'.replace("\\", "/")
                
                
                
                #https://v.redd.it/x2hinx0binu21
        elif "v.redd.it" in url:
            # This is a v.redd.it post
            file_path = os.path.join(f"{output_dir}/{name}", f'{safe_title}.mp4')
            if not os.path.exists(file_path):
                command = f"youtube-dl {url} -o \"{file_path}\""
                process = subprocess.Popen(command, shell=False)
                process.wait()
                
            #images_data.append({'url': f"{output_dir}/{name}/{safe_title}.mp4".replace("\\", "/"), 'title': safe_title})
            localurl = f"{output_dir}/{name}/{safe_title}.mp4".replace("\\", "/")
            
                
                #https://i.imgur.com/mQlSXHO.gifv
                #https://imgur.com/download/mQlSXHO
        elif "imgur.com" in url and url.endswith(".gifv"):
            # This is an imgur post
            file_name = f'{safe_title}.mp4'
            file_direct = f"{output_dir}/{name}"
            file_path = os.path.join(f"{file_direct}", f'{file_name}') 
            
            if not os.path.exists(file_name):
                command = f"gallery-dl {url} -D \"{file_direct}\" -f \"{file_name}\""

                process = subprocess.Popen(command, shell=True)
                process.wait()
                
              
            #images_data.append({'url': file_path.replace("\\", "/"), 'title': safe_title})   
            localurl = file_path.replace("\\", "/")    
                        
        elif "imgur.com" in url and (url.endswith(".jpg") or url.endswith(".png")):
            # This is an imgur post
            file_name = f'{safe_title}.jpg'
            file_direct = f"{output_dir}/{name}"
            file_path = os.path.join(f"{file_direct}", f'{file_name}') 
            
            if not os.path.exists(file_name):
                command = f"gallery-dl {url} -D \"{file_direct}\" -f \"{file_name}\""

                process = subprocess.Popen(command, shell=False)
                process.wait()
              
            #images_data.append({'url': file_path.replace("\\", "/"), 'title': safe_title})
            localurl = file_path.replace("\\", "/")
            
        else:
            # This is a single image post
            file_extension = os.path.splitext(url.split("?")[0])[1]
            if file_extension == "":
                return None
                
            file_path = os.path.join(f"{output_dir}/{name}", f'{safe_title}{file_extension}')
            if not os.path.exists(file_path):
                response = requests.get(url)
                with open(file_path, 'wb') as f:
                    f.write(response.content)
                    
            #images_data.append({'url': f"{output_dir}/{name}/{safe_title}{file_extension}".replace("\\", "/"), 'title': safe_title})
            localurl = f"{output_dir}/{name}/{safe_title}{file_extension}".replace("\\", "/")
    except:
        traceback.print_exc()
        return None
    return localurl


def postedsincecalc(post):
    time_difference = (datetime.now(timezone.utc) - datetime.fromtimestamp(post.created_utc, timezone.utc))
    time_since_post = [time_difference.seconds,
                        time_difference.seconds // 60,
                        time_difference.seconds // 3600,
                        time_difference.days,
                        time_difference.days // 30,
                        time_difference.days // 365]

    postedsince = "Unknown"
    
    for i in range(len(time_since_post)):
        if time_since_post[i] == 0:
            
            postedsince = str(time_since_post[i-1]) + " " + ["sec.", "min.", "hr.", "days", "mo.", "yr."][i-1] + " ago"
            return postedsince
        
    return str(time_since_post[-1]) + " yr. ago"

        
def download_images(subreddit_name, output_dir,limit=10,post_type="top",since="all"):
    #print(subreddit_name, output_dir)
    reddit = praw.Reddit(
        client_id = os.getenv('praw_api_client_id'),
        client_secret = os.getenv('praw_api_client_secret'),
        user_agent = "testscript by RS_ted ",
        redirect_uri = "http://localhost:5353"
    )

    name = subreddit_name

    subreddit = reddit.subreddit(name)
    #print(subreddit.top(limit=100))
    if post_type == "top":
        top_posts = subreddit.top(time_filter=since,limit=limit)
    elif post_type == "hot":
        top_posts = subreddit.hot(limit=limit)
    elif post_type == "new":
        top_posts = subreddit.new(limit=limit)
    elif post_type == "rising":
        top_posts = subreddit.rising(limit=limit)

    #input(f"Downloading {name} images to {output_dir}. Press Enter to continue...")
    # Create the directory if it doesn't exist
    if not os.path.exists(f"{output_dir}/{name}"):
        os.makedirs(f"{output_dir}/{name}")
 
    images_data = []
    #C:/Users/rouna/OneDrive/Desktop/Nodejs_intro/redditdownlaoder/reddit_gallery_project/reddit_gallery/media/subreddit_name
    for post in (top_posts):
        try:
            try:
                url = post.url
            except Exception as e:
                print(f"Error: {e}")
                url=None

            try:
                title = post.title
            except Exception as e:
                print(f"Error: {e}")

                title=None

            try:
                author = str(post.author)
            except Exception as e:
                print(f"Error: {e}")
     
                author=None

            try:
                authordp = post.author.icon_img
            except Exception as e:
                print(f"Error: {e}")
                authordp="https://i.redd.it/snoovatar/avatars/8658e16c-55fa-486f-b7c7-00726de2e742.png"

            try:
                postedsince = postedsincecalc(post)
            except Exception as e:
                print(f"Error: {e}")
 
                postedsince=None
            
            try:
                flair = str(post.link_flair_text)
            except:
                flair = "none"
            
            try:
                comments = f"{int(post.num_comments) // 1000}k" if int(post.num_comments) >= 1000 else int(post.num_comments)
            except Exception as e:
                print(f"Error: {e}")

                comments='NaN'

            try:
                upvotes = f"{int(post.score) // 1000}k" if int(post.score) >= 1000 else int(post.score)
            except Exception as e:
                print(f"Error: {e}")

                upvotes='NaN'

            try:
                downvotes = f"{int(post.score*(1-post.upvote_ratio)) // 1000}k" if int(post.score*(1-post.upvote_ratio)) >= 1000 else int(post.score*(1-post.upvote_ratio))
                #print("downvotes: ",post.upvote_ratio, post.score*(1-post.upvote_ratio))
            except Exception as e:
                print(f"Error: {e}")

                downvotes='NaN'

            try:
                safe_title = ''.join(c for c in f"{title} by {author}" if c.isalnum() or c in (' ',)).rstrip()
            except Exception as e:
                print(f"Error: {e}")

                safe_title=None
            localurl = urltolocalurl(url,post,name,output_dir,safe_title)
            if localurl == None:
                continue
            siteurl = "https://www.reddit.com"+str(post.permalink)
            userurl = "https://www.reddit.com/user/"+str(post.author)


            images_data.append({'url': localurl, 'title': safe_title, 'author': author,
                                'author_dp': authordp, 'posted_since': postedsince,
                                'flair': flair, 'comments': comments, 'upvotes': upvotes,
                                'downvotes': downvotes,'siteurl':siteurl,'userurl':userurl,'mediaurl':url})    
            
              
            print("["+json.dumps(images_data[-1])+"]")   
            
        except Exception as e:
            print(f"Error: {e}")
            traceback.print_exc()
            continue
 

#download_images("nsfw_gifs", "public\media", 10, "top", "all")

if __name__ == "__main__":

    # Command line arguments: subreddit name and output directory
    subreddit_name = sys.argv[1]
    output_dir = sys.argv[2] if len(sys.argv) > 2 else "public\media"
    limit = int(sys.argv[3]) if len(sys.argv) > 3 else 10
    post_type = sys.argv[4] if len(sys.argv) > 4 else "top"
    since = sys.argv[5] if len(sys.argv) > 5 else "all"

    download_images(subreddit_name, output_dir, limit, post_type, since)




# download_images(input("enter subrrdiit: "), "public\media")

# python downloader.py "swimsuitsuccubus" "public\media" 10 top all

File: downloader.js
import snoowrap from 'snoowrap';
import axios from 'axios';
import fs from 'fs';
import path from 'path';
import { exec } from 'child_process';
import { toZonedTime } from 'date-fns-tz';
import ora from 'ora';
import util from 'util';
import {getGif} from 'redgif';
import process from 'node:process';
import e from 'express';
import qs from 'qs';

async function getAccessToken() {
    const authString = `${process.env.praw_api_client_id}:${process.env.praw_api_client_secret}`;
    const authBuffer = Buffer.from(authString, 'utf8');
    const authBase64 = authBuffer.toString('base64');
  
    const response = await axios.post('https://www.reddit.com/api/v1/access_token', qs.stringify({
      grant_type: 'client_credentials',
    }), {
      headers: {
        'Authorization': `Basic ${authBase64}`,
        'Content-Type': 'application/x-www-form-urlencoded',
      },
    });
  
    const r = new snoowrap({
        userAgent: 'testscript by RS_ted',
        clientId: process.env.praw_api_client_id,
        clientSecret: process.env.praw_api_client_secret,
        accessToken: response.data.access_token,
        });

    console.log("Access Token: ", r.accessToken);

    return r;
}

async function urltolocalurl(url, post, name, outputDir, safeTitle) {
    try {

        console.log("url: ", url);
        if (url.includes('reddit.com/gallery')) {
            // This is a gallery post

            const localUrls = [];
            for (let key in post.media_metadata) {
                //console.log("key: ", key);
                const media = post.media_metadata[key];
                if (media.e === "Image" || media.e === "AnimatedImage") {
                    let mediaUrl;
                    mediaUrl = media.s.gif ? media.s.gif : media.s.u;
                    //console.log("mediaUrl: ", mediaUrl);
                    const fileExtension = path.extname(mediaUrl.split('?')[0]);
                    const filePath = path.join(outputDir, `${safeTitle}_${key}${fileExtension}`);
                    if (!fs.existsSync(filePath)) {
                        const response = await axios.get(mediaUrl, { responseType: 'stream' });
                        const writer = fs.createWriteStream(filePath);
                        response.data.pipe(writer);
                        new Promise((resolve, reject) => {
                            writer.on('finish', resolve);
                            writer.on('error', reject);

                        localUrls.push(mediaUrl);
                            
                    });
                    }else{
                    localUrls.push(filePath.replace(/\\/g, '/'));
                    }
                    //console.log(`Downloaded ${filePath.replace(/\\/g, '/')}`);
                }
            }
            return localUrls;
        } else if (url.includes('redgifs.com/watch')) {
            // This is a redgifs vid
            const filePath = path.join(outputDir, `${safeTitle}.mp4`);
            if (!fs.existsSync(filePath)) {
                const gifId = url.split('/').pop().split('#')[0];
                const buffer = await getGif(gifId);
                const writer = fs.writeFileSync(filePath, buffer)
                
                new Promise((resolve, reject) => {
                    fs.writeFile(filePath, buffer, (err) => {
                        if (err) {
                            reject(err);
                        } else {
                            resolve();
                        }
                    });
                });
                return filePath.replace(/\\/g, '/');
            }
            else {
            return filePath.replace(/\\/g, '/');
            }
        }
        else if (url.includes('i.redgifs')) {
            // This is a redgifs image
            const filePath = path.join(outputDir, `${safeTitle}.jpg`);
            if (!fs.existsSync(filePath)) {
                const gifId = url.split('/').pop().split('#')[0].split('.')[0];
                const buffer = await getGif(gifId);
                

                new Promise((resolve, reject) => {
                    fs.writeFile(filePath, buffer, (err) => {
                        if (err) {
                            reject(err);
                        } else {
                            resolve();
                        }
                    });
                });

                return filePath.replace(/\\/g, '/');
            }
            return filePath.replace(/\\/g, '/');
        }

        else if (url.includes('v.redd.it')) {
            // This is a v.redd.it post
            //console.log("!!!This is a v.redd.it post!!!");
            const filePath = path.join(outputDir, `${safeTitle}.mp4`);
            if (!fs.existsSync(filePath)) {
                //console.log("!!!Debugging!!!   ", post);
                
                const child = exec(`youtube-dl ${url} -o "${filePath}"`);
                new Promise((resolve, reject) => {
                    child.on('close', resolve);
                    child.on('error', reject);
                });

                return post.secure_media.reddit_video.hls_url;
            }
            else {  
            return filePath.replace(/\\/g, '/');
            }
        } else if (url.includes('imgur.com') && url.endsWith('.gifv')) {
            // This is an imgur post
            const fileName = `${safeTitle}.mp4`;
            const fileDir = path.join(outputDir);
            const filePath = path.join(fileDir, fileName);
            if (!fs.existsSync(filePath)) {
                const child = exec(`gallery-dl ${url} -D "${fileDir}" -f "${fileName}"`);
                await new Promise((resolve, reject) => {
                    child.on('close', resolve);
                    child.on('error', reject);
                });
            }
            return filePath.replace(/\\/g, '/');
        } else if (url.includes('imgur.com') && (url.endsWith('.jpg') || url.endsWith('.png'))) {
            // This is an imgur post
            const fileName = `${safeTitle}.jpg`;
            const fileDir = path.join(outputDir);
            const filePath = path.join(fileDir, fileName);
            if (!fs.existsSync(filePath)) {
                const child = exec(`gallery-dl ${url} -D "${fileDir}" -f "${fileName}"`);
                await new Promise((resolve, reject) => {
                    child.on('close', resolve);
                    child.on('error', reject);
                });
            }
            return filePath.replace(/\\/g, '/');
        } else {
            // This is a single image post
            const fileExtension = path.extname(url.split('?')[0]);
            if (!fileExtension) return url;
            const filePath = path.join(outputDir, `${safeTitle}${fileExtension}`);
            if (!fs.existsSync(filePath)) {
                const response = await axios.get(url, { responseType: 'stream' });
                console.log("response: ", response);
                const writer = fs.createWriteStream(filePath);
                response.data.pipe(writer);
                new Promise((resolve, reject) => {
                    writer.on('finish', resolve);
                    writer.on('error', reject);
                });
                return url;
            }
            else {
                return filePath.replace(/\\/g, '/');
            }
        }
    } catch {
        console.log("Error in urltolocalurl");
        return url;
    }
}

function postedSinceCalc(post) {
    const units = [
        { name: "yr.", ms: 12 * 30 * 24 * 60 * 60 * 1000 },
        { name: "mo.", ms: 30 * 24 * 60 * 60 * 1000 },
        { name: "day", ms: 24 * 60 * 60 * 1000 },
        { name: "hr.", ms: 60 * 60 * 1000 },
        { name: "min.", ms: 60 * 1000 },
        { name: "sec.", ms: 1000 }
    ];

    let timeDifference = Date.now() - (post.created_utc * 1000);

    for (let unit of units) {
        const value = Math.floor(timeDifference / unit.ms);
        if (value > 0) {
            return `${value} ${unit.name} ago`;
        }
    }
    return "just now";
}

export async function* downloadImages(subredditName, outputDir = './public/media', limit = 10, postType = 'top', since = 'all',r = null, signal = null) {

    if (!r){
        r = await getAccessToken();
        console.log("guest mode");
    } else
    {
        console.log("user mode");
    }

    const snooWrap = new snoowrap({
        userAgent: r.userAgent,
        clientId: r.clientId,
        clientSecret: r.clientSecret,
        refreshToken: r.refreshToken,
        accessToken: r.accessToken
    });

    r = snooWrap;
    
    console.log("r: ", r);

    const spinner = ora('Fetching posts...').start();
    let name = subredditName;

    let subreddit;
    let Posts;
    if (name === ''){
        Posts = await r.getHot();
        name = '_other_';
    }
    else{

        subreddit = await r.getSubreddit(name);
        switch (postType) {
            case 'top':
                Posts = await subreddit.getTop({ time: since, limit });
                break;
            case 'hot':
                Posts = await subreddit.getHot({ limit });
                break;
            case 'new':
                Posts = await subreddit.getNew({ limit });
                break;
            case 'rising':
                Posts = await subreddit.getRising({ limit });
                break;
            default:
                Posts = await subreddit.getTop({ time: since, limit });
        }
    }

    spinner.stop();

    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }


    const outputPath = outputDir + '/' + name;
    if (!fs.existsSync(outputPath)) {
        fs.mkdirSync(outputPath, { recursive: false });
    }

    for (const post of Posts) {

        if (signal && signal.aborted) {
            console.log('Generator aborted.');
            return;
          }

        let url;

        try {
            url = post.url;
            let aspectRatio;
            if (post.is_gallery) {
                const galleryData = post.media_metadata;
                let maxAspectRatio = -1;
                for (let key in galleryData) {
                    const media = galleryData[key];
                    if (media.e === "Image" || media.e === "AnimatedImage") {
                        const aspectRatio = media.s.x/media.s.y;
                        if (aspectRatio > maxAspectRatio) {
                            maxAspectRatio = aspectRatio;
                        }
                    }
                }
                aspectRatio = maxAspectRatio;
            }

            else if (post.is_video && post.is_reddit_media_domain == true) {
                aspectRatio = post.secure_media.reddit_video.width / post.secure_media.reddit_video.height;
            }

            
            
            else if (post.is_reddit_media_domain == false && post.preview && post.preview.reddit_video_preview != null) {

                aspectRatio = post.preview.reddit_video_preview.width / post.preview.reddit_video_preview.height;

            }

            
            else if (post.preview && post.preview.images[0].source != null) {
                try {
                    aspectRatio = post.preview.images[0].source.width / post.preview.images[0].source.height;
                } catch {
                    aspectRatio = NaN;
                }
            }

        
            const title = post.title;
            const author = post.author.name;
            let authorDp;
            try {
                authorDp = await post.author.icon_img;
                if (!authorDp) {
                    authorDp = 'https://i.redd.it/snoovatar/avatars/8658e16c-55fa-486f-b7c7-00726de2e742.png';
                }
            } catch {
                authorDp = 'https://i.redd.it/snoovatar/avatars/8658e16c-55fa-486f-b7c7-00726de2e742.png';
            }
            const postedSince = postedSinceCalc(post);
            const flair = post.link_flair_text || 'none';
            const comments = post.num_comments >= 1000 ? `${Math.floor(post.num_comments / 1000)}k` : post.num_comments.toString();
            const upvotes = post.score >= 1000 ? `${Math.floor(post.score / 1000)}k` : post.score.toString();
            const downvotes = Math.floor(post.score * (1 - post.upvote_ratio)) >= 1000 ? `${Math.floor(post.score * (1 - post.upvote_ratio) / 1000)}k` : Math.floor(post.score * (1 - post.upvote_ratio)).toString();
            const isNSFW = post.over_18;
            const safeTitle = `${title} by ${author}`.replace(/[^a-zA-Z0-9 ]/g, '').trim();

            // const fileExists = fs.existsSync(path.join(outputPath, `${safeTitle}.*`));
    
                const localUrl = await urltolocalurl(url, post, name, outputPath, safeTitle);
                if (localUrl) {
                    const imageData = {
                        url: localUrl,
                        aspect_ratio: aspectRatio,
                        title: title,
                        author,
                        author_dp: authorDp,
                        posted_since: postedSince,
                        flair,
                        comments,
                        upvotes,
                        downvotes,
                        siteurl: `https://www.reddit.com${post.permalink}`,
                        userurl: `https://www.reddit.com/user/${author}`,
                        mediaurl: url,
                        isNSFW: isNSFW
                    };
                    yield imageData;
                    //console.log(JSON.stringify(imageData));
                }

        } catch (err) {
            
            console.error("Error in url:",url);
            console.error(err);
            yield { error: err };
            console.log("Error in url:",url);
            console.log('Error :' + err);
        }
    }

}



// export async function* downloadImages(subredditName, outputDir, limit = 10, postType = 'top', since = 'all') {
//     let spinner = ora('Fetching posts...').start();
//     const name = subredditName;

//     let url;
//     switch (postType) {
//         case 'top':
//             url = `https://www.reddit.com/r/${name}/top.json?t=${since}&limit=${limit}`;
//             break;
//         case 'hot':
//             url = `https://www.reddit.com/r/${name}/hot.json?limit=${limit}`;
//             break;
//         case 'new':
//             url = `https://www.reddit.com/r/${name}/new.json?limit=${limit}`;
//             break;
//         case 'rising':
//             url = `https://www.reddit.com/r/${name}/rising.json?limit=${limit}`;
//             break;
//         default:
//             url = `https://www.reddit.com/r/${name}/top.json?t=${since}&limit=${limit}`;
//     }

//     console.log("url: ", url);

//     let response;
//     try {
//         response = await axios.get(url);
//         //console.log(response);
//     } catch (error) {
//         console.error(`Error fetching posts: ${error}`);
//         return;
//     }
//     spinner.stop();

//     const outputPath = outputDir + '/' + name;
//     if (!fs.existsSync(outputPath)) {
//         fs.mkdirSync(outputPath, { recursive: false });
//     }

//     const Posts = response.data.data.children.map(child => child.data);
//     let authorResponse;
//     for(const post of Posts) {

//         spinner = ora(`Downloading ${post.title}...`).start();

//         try {

//             const url = post.url;
//             const title = post.title;
//             const author = post.author;
//             let authorDp;
//             if (post.author != '[deleted]') {
//                 try {
//                     authorResponse = await axios.get(`https://www.reddit.com/user/${post.author}/about.json`);
//                     //console.log("authorResponse: ", authorResponse);
//                     authorDp = authorResponse.data.data.icon_img.split('?')[0];

//                 } catch (error) {
//                     authorDp = 'https://i.redd.it/snoovatar/avatars/8658e16c-55fa-486f-b7c7-00726de2e742.png';
//                 }
//             }else {
//                 authorDp = 'https://i.redd.it/snoovatar/avatars/8658e16c-55fa-486f-b7c7-00726de2e742.png';
//             }
//             const postedSince = postedSinceCalc(post);
//             const flair = post.link_flair_text || 'none';
//             const comments = post.num_comments >= 1000 ? `${Math.floor(post.num_comments / 1000)}k` : post.num_comments.toString();
//             const upvotes = post.score >= 1000 ? `${Math.floor(post.score / 1000)}k` : post.score.toString();
//             const downvotes = Math.floor(post.score * (1 - post.upvote_ratio)) >= 1000 ? `${Math.floor(post.score * (1 - post.upvote_ratio) / 1000)}k` : Math.floor(post.score * (1 - post.upvote_ratio)).toString();
//             const isNSFW = post.over_18;
//             const safeTitle = `${title} by ${author}`.replace(/[^a-zA-Z0-9 ]/g, '').trim();

//             const fileExists = fs.existsSync(path.join(outputPath, `${safeTitle}.*`));

//             const localUrl = await urltolocalurl(url, post, name, outputPath, safeTitle);

//             if (localUrl) {
//                 const imageData = {
//                     url: Array.isArray(localUrl) ? localUrl : localUrl,
//                     title: title,
//                     author,
//                     author_dp: authorDp,
//                     posted_since: postedSince,
//                     flair,
//                     comments,
//                     upvotes,
//                     downvotes,
//                     siteurl: `https://www.reddit.com${post.permalink}`,
//                     userurl: `https://www.reddit.com/user/${author}`,
//                     mediaurl: url,
//                     isNSFW: isNSFW
//                 };
//                 //console.log(JSON.stringify(imageData));
//                 yield imageData;
//             }
//         }
//         catch (err) {
//             console.error(err);
//             yield { error: err };
//             console.log('Error :' + err);
//         }

//         spinner.stop();
//     } 
// }


// if (process.argv.length > 2) {
//     const [, , subredditName, outputDir = 'public/media', limit = 10, postType = 'top', since = 'all'] = process.argv;
//     downloadImages(subredditName, outputDir, limit, postType, since);
//   }

// const imageGenerator = downloadImages('news', './public/media', 2, 'top', 'today');

// for await (const value of imageGenerator) {
//     if (value instanceof Promise) {
//         await value; // Wait for the download promise to resolve
//     } else {
//         // Do something with the imageData object
//         console.log(value);
//     }
// }